src/postgraphile/http/subscriptions.ts

```diff
@@ -186,134 +191,262 @@ export async function enhanceHttpServerWithSubscriptions<
+              /*
+               * Enable JWT support through connectionParams.
+               *
+               * For other headers you'll need to do this yourself for security
+               * reasons (e.g. we don't want to allow overriding of Origin /
+               * Referer / etc)
+               */
+              request.headers.authorization = String(normalizedConnectionParams['authorization']);
+            }
+
+            socket['postgraphileHeaders'] = {
+              ...normalizedConnectionParams,
+              // The original headers must win (for security)
+              ...request.headers,
+            };
+          },
+          async onSubscribe(ctx, msg) {
+            const context = await getContext(ctx.socket, msg.id);
+
+            // Override schema (for --watch)
+            const schema = await getGraphQLSchema();
+
+            const { payload } = msg;
+            const args: ExecutionArgs = {
+              schema,
+              contextValue: context,
+              operationName: payload.operationName,
+              document: parse(payload.query),
+              variableValues: payload.variables,
+            };
+
+            // for supplying custom execution arguments
+            const hookedArgs = pluginHook
+              ? pluginHook('postgraphile:ws:onSubscribe', args, {
+                  ...ctx,
+                  message: msg,
+                  options,
+                })
+              : args;
+
+            // when supplying custom execution args from the
+            // onSubscribe, you're trusted to do the validation
+            let validationErrors = validate(
+              hookedArgs.schema,
+              hookedArgs.document,
+              staticValidationRules,
             );
-            error['errors'] = validationErrors;
-            return Promise.reject(error);
-          }
-        }
+            if (validationErrors.length) {
+              return validationErrors;
+            }
+
+            // You are strongly encouraged to use
+            // `postgraphile:validationRules:static` if possible - you should
+            // only use this one if you need access to variables.
+            const { req, res } = await reqResFromSocket(ctx.socket);
+            const validationRules = pluginHook('postgraphile:validationRules', [], {
+              options,
+              req,
+              res,
+              variables: hookedArgs.variableValues,
+              operationName: hookedArgs.operationName,
+              // no meta because validation errors returned from here will be
+              // served through the error message. it contains just the GraphQLErrors
+              // (there is no result to add the meta to)
+            });
+            if (validationRules.length) {
+              validationErrors = validate(hookedArgs.schema, hookedArgs.document, validationRules);
+              if (validationErrors.length) {
+                return validationErrors;
```

Comment:
I prefer that `const` is used where possible; it makes the code easier for me to follow because I don't have to worry about rebinding.

```suggestion
            const moreValidationRules = pluginHook('postgraphile:validationRules', [], {
              options,
              req,
              res,
              variables: hookedArgs.variableValues,
              operationName: hookedArgs.operationName,
              // no meta because validation errors returned from here will be
              // served through the error message. it contains just the GraphQLErrors
              // (there is no result to add the meta to)
            });
            if (moreValidationRules.length) {
              const moreValidationErrors = validate(hookedArgs.schema, hookedArgs.document, moreValidationRules);
              if (moreValidationErrors.length) {
                return moreValidationErrors;
```